### DISCUSSION TITLE
Does etcd-server read from disk when sending snapshots to followers

### DISCUSSION THREAD

**Discussion by Zhiying12** [etcd-io/etcd#16514] [Q&A] [answered]
Given a cluster of three nodes, let's say we start two of them at the beginning.  After loading several GiB keys, we start the third one. Based on the network monitoring log, I can see that leader sends its db to the new peer as a snapshot. However, I also monitored disk I/O, and I couldn't see the leader node read anything from the disk. 

Does the leader maintain the whole db in memory?

**Comment by fuweid** [comment, marked as answer]
Hi, did you check the page cache in the leader?

The leader reads the boltdb from disk and send it to follower. The boltdb is mmaped into memory.
If the file has page cache in kernel, I think there is not disk IO.

REF: https://github.com/etcd-io/etcd/blob/60051be9908649b1a0f2d000dc75b3bd0822d53c/server/etcdserver/snapshot_merge.go#L38

**Reply by Zhiying12** [reply]
Appreciate it! A follow-up question. Does etcd do regular snapshot for v3 store (boltdb)? I checked the code for snapshot, and it seems it only does snapshot for v2 store. How does a follower recover the state from the snapshot sent by the leader?

**Reply by ahrtr** [reply]
> Does etcd do regular snapshot for v3 store (boltdb)?

No. The leader only sends a snapshot to a slow follower when it's far lag behind the leader's data.

> How does a follower recover the state from the snapshot sent by the leader?

Pls read 
https://github.com/etcd-io/etcd/blob/27633923e45bca9e6699c9d68df5790840370728/server/etcdserver/server.go#L941

**Reply by Zhiying12** [reply]
Thanks a lot! IMHO, the follower will replace its backend db with the leader's snapshot below.
https://github.com/etcd-io/etcd/blob/27633923e45bca9e6699c9d68df5790840370728/server/storage/backend.go#L59-L66
After updating the db, the follower restores the **corresponding** `consistIndex` from the snapshot. So far, the follower catches up all the progress up to the snapshot. After that, it just needs to apply entries after the consistIndex. Is my understanding correct?
https://github.com/etcd-io/etcd/blob/27633923e45bca9e6699c9d68df5790840370728/server/etcdserver/server.go#L989-L990

Also, in `snapshot()` below, the etcd server just takes snapshot of V2 store and trims outdated log entries. It would creates several v2 snapshot files with suffix ".snap" under the snap folder. As v2 store is deprecated, does it just keep track of metadata, or are these files still useful other than compatibility purpose?
https://github.com/etcd-io/etcd/blob/27633923e45bca9e6699c9d68df5790840370728/server/etcdserver/server.go#L2066

Thanks everyone for your time and patience again!

**Reply by fuweid** [reply]
> After that, it just needs to apply entries after the consistIndex. Is my understanding correct?

I didn't follow your question. After restore the snapshot, the leader can sync the log with follower. And then the follower applies the log.

> As v2 store is deprecated, does it just keep track of metadata, or are these files still useful other than compatibility purpose?

The deprecated work is still WIP. Hope the following doc can help you.

* https://github.com/etcd-io/etcd/pull/15754
* https://github.com/etcd-io/etcd/issues/12913

**Reply by Zhiying12** [reply]
Thank you so much for your answer and reference!
